### PR TITLE
HubClient Implementation for WebSub Hub

### PR DESCRIPTION
**Pull request by ayeshLK**

### CODE REVIEW

**Comment by niveathika** [comment]
@ayeshLK Please go through and follow GitHub commit message best practises for future PRs, https://chris.beams.io/posts/git-commit.

File: websubhub-ballerina/commons.bal [hunk 1]
```diff
@@ -164,3 +172,11 @@ public readonly class StatusPermanentRedirect {
...
 
 final StatusPermanentRedirect STATUS_PERMANENT_REDIRECT = new;
+
+isolated function isSuccessStatusCode(int statusCode) returns boolean {
+    return (200 <= statusCode && statusCode < 300);
+}
+
+isolated function generateLinkUrl(string hubUrl, string topic) returns string {
+    return hubUrl + "; rel=\"hub\", " + topic + "; rel=\"self\"";
+}
```
Comment (niveathika) [inline websubhub-ballerina/commons.bal:182]:
```suggestion
}

```

File: websubhub-ballerina/hub_client.bal [hunk 2]
```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2018 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (niveathika) [inline websubhub-ballerina/hub_client.bal:1]: License year should be 2021

File: websubhub-ballerina/hub_client.bal [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        check request.setContentType(contentType);
+        
+        if (msg?.headers is map<string|string[]>) {
+            var headers = <map<string|string[]>>msg?.headers;
+            foreach var [header, values] in headers.entries() {
+                if (values is string) {
+                    request.addHeader(header, values);
+                } else {
+                    string headerValue = ";".'join(...<string[]>values) + ";";
+                    request.addHeader(header, headerValue);
```
Comment (niveathika) [inline websubhub-ballerina/hub_client.bal:77]: add header actually appends the header values together, you can simply call addHeader() method inside a for loop

File: websubhub-ballerina/tests/basic_hub_test.bal [hunk 4]
```diff
@@ -23,7 +23,7 @@ boolean isValidationFailed = false;
 
 http:Client httpClient = checkpanic new("http://localhost:9090/websubhub");
 
-listener Listener functionWithArgumentsListener = new(9090);
+listener Listener functionWithArgumentsListener = new(9090, "https://sample.hub.com");
```
Comment (niveathika) [inline websubhub-ballerina/tests/basic_hub_test.bal:26]: [URGENT] Do we have to ask this explicitly? Wouldnt we be able to get this from the underlying http client?

File: websubhub-ballerina/commons.bal [hunk 5]
```diff
@@ -49,16 +47,25 @@ const string MODE_UNSUBSCRIBE = "unsubscribe";
...
-    string|xml|json|byte[]|io:ReadableByteChannel payload = "";
-    string contentType = "";
+# + content - The payload to be sent
+public type ContentDistributionMessage record {|
+    map<string|string[]>? headers = ();
+    string? contentType = ();
+    json|xml|string|byte[] content;
+|};
+
+public type ContentDistributionSuccess record {|
```
Comment (shafreenAnfar) [inline websubhub-ballerina/commons.bal:70]: I think need to change this type.

It should be something like the below

```ballerina
const int STATUS_OK = 200

type Status distinct object {
    public int code;
};

public readonly class StatusOK {
    *Status;
    public STATUS_OK code = STATUS_OK;
}

final StatusOK STATUS_OK_OBJ = new;

type CommonResponse record {|
    string mediaType?;
    map<string|string[]> headers?;
    anydata body?;
|};

public type ContentDistributionSuccess record {|
    *CommonResponse;
    readonly StatusOK status = STATUS_OK_OBJ;
|};
```

**Review by shafreenAnfar — APPROVED** [review]

**Review by niveathika — APPROVED** [review]

**Pull request merged**